### PR TITLE
initial keras-hub support

### PR DESCRIPTION
**Pull request by martin-gorner**
Initial support for Keras-hub:

- generic snippet for loading models
- library definition with metadata

### CODE REVIEW

File: packages/tasks/src/model-libraries-snippets.ts
Comment (julien-c) [inline packages/tasks/src/model-libraries-snippets.ts:368]: let's just change the deafault to jax but not reformat?

Comment (martin-gorner) [reply packages/tasks/src/model-libraries-snippets.ts:368]: sure

File: packages/tasks/src/model-libraries-snippets.ts
Comment (julien-c) [inline packages/tasks/src/model-libraries-snippets.ts:400]: do you know if there's a way to know which task/architecture to use, depending on what's inside the repo? (or a manual tag set by keras.push?)

File: packages/tasks/src/model-libraries-snippets.ts
Comment (julien-c) [inline packages/tasks/src/model-libraries-snippets.ts:404]: maybe not worth adding, maybe just the main arch/task is better/simpler

Comment (martin-gorner) [reply packages/tasks/src/model-libraries-snippets.ts:404]: sure, that works too.

**Comment by martin-gorner** [comment]
Not sure about the tests failing. It does not look like it's related to this PR, is it?

File: packages/tasks/src/model-libraries-snippets.ts
Comment (Wauplin) [reply packages/tasks/src/model-libraries-snippets.ts:400]: looking at https://huggingface.co/martin-gorner/llama-3.2-1B-pirate-instruct, it seems the architecture is pushed with the model card:

> This model is related to a CausalLM task.

So yeah agree with Julien that if we can have this info in the metadata, it would be awesome so simplify the snippet

**Comment by Wauplin** [comment]
(failing tests don't seem related no)

File: packages/tasks/src/model-libraries-snippets.ts
Comment (martin-gorner) [reply packages/tasks/src/model-libraries-snippets.ts:400]: I'm currently investigating this. Yes, this would be the best solution. I do see a text-generation "pipeline-tag" on models that I upload. This is probably what we should use. It won't be part of this PR though.

Comment (martin-gorner) [reply packages/tasks/src/model-libraries-snippets.ts:400]: 😫 right now the pipeline-tag is only populated for CausalLM and TextClassifier models.
https://github.com/keras-team/keras-hub/blob/a45110ee4f04750cf43214b591847cffa1074310/keras_hub/src/utils/preset_utils.py#L320

Comment (martin-gorner) [reply packages/tasks/src/model-libraries-snippets.ts:400]: btw, the "example use" section on Kaggle is much more complete. For reference: https://www.kaggle.com/models/keras/gemma2

Comment (Wauplin) [reply packages/tasks/src/model-libraries-snippets.ts:400]: > btw, the "example use" section on Kaggle is much more complete. For reference: https://www.kaggle.com/models/keras/gemma2

I feel that this kind of completeness should be written in the model card directly. For example, https://huggingface.co/google/gemma-2-9b-keras model card could contain code snippets on how to generate and fit on data. What we try to achieve in "use this model" is to have a small, self-contained and personalized snippet to load a model

Comment (julien-c) [reply packages/tasks/src/model-libraries-snippets.ts:400]: agree

Comment (Vaibhavs10) [reply packages/tasks/src/model-libraries-snippets.ts:400]: +1 on making the snippet more complete; in general, we've been pushing for more complete, end-to-end snippets that people can just copy and run.

File: packages/tasks/src/model-libraries-snippets.ts
Comment (pcuenca) [inline packages/tasks/src/model-libraries-snippets.ts:397]:
```suggestion
model = keras_hub.models.CausalLM.from_preset("hf://${model.id}", dtype="bfloat16")
```

File: packages/tasks/src/model-libraries-snippets.ts
Comment (Wauplin) [inline packages/tasks/src/model-libraries-snippets.ts:400]: what do you think of having a method that does something along the lines:

```ts
export const kerasHub = (model: ModelData): string[] | undefined => {
    let kerasCls: string | undefined;

    if (model.pipeline_tag === "text-classification") {
        kerasCls = "TextClassifier";
    } else if (model.pipeline_tag === "text-generation") {
        kerasCls = "CausalLM";
    }

    if (!kerasCls) {
        return undefined;
    }

    return [
        `# Available backend options are: "jax", "torch", "tensorflow".
import os
os.environ["KERAS_BACKEND"] = "jax"

import keras_hub

model = keras_hub.models.${kerasCls}.from_preset("hf://${model.id}", dtype="bfloat16")
`,
    ];
};
```

In the future, we can edit the snippet to support more architectures.

Comment (Wauplin) [reply packages/tasks/src/model-libraries-snippets.ts:400]: this way if we know how to instantiate a model, a snippet will be shown. If not, we don't show anything and the user has to read Keras Hub docs (which seems to be the case anyway).

Comment (martin-gorner) [reply packages/tasks/src/model-libraries-snippets.ts:400]: I'd rather parse task.json and generate the snippet with the correct class directly instead of trying to infer it from the pipeline-tag. The info is there.

Comment (martin-gorner) [reply packages/tasks/src/model-libraries-snippets.ts:400]: But that is a slightly larger PR and I'd like to do it separately

Comment (Wauplin) [reply packages/tasks/src/model-libraries-snippets.ts:400]: > I'd rather parse task.json and generate the snippet with the correct class directly instead of trying to infer it from the pipeline-tag. The info is there.

That's even better yes!

> But that is a slightly larger PR and I'd like to do it separately

Makes sense. Just a nit but tn that case, I'd not add a snippet at all (snippets are optional anyway). This way we won't have  "half-usable snippets" until the PR to parse task.json is deployed.

**Review by Wauplin — APPROVED** [review]
Looks good to merge! :)

To sum-up, follow-up PRs will be about 1. parsing task.json for better code snippets 2. removing `keras-nlp` library entirely (in a few weeks?).

Failing CI doesn't seem related to this PR. Let's wait for another approval given the discussions in this PR :)

File: packages/tasks/src/model-libraries.ts
Comment (julien-c) [inline packages/tasks/src/model-libraries.ts:357]:
```suggestion
```

false is the default

File: packages/tasks/src/model-libraries.ts
Comment (pcuenca) [inline packages/tasks/src/model-libraries.ts:348]: Same as below (remove the line)

**Review by pcuenca — APPROVED** [review]

File: packages/tasks/src/model-libraries.ts
Comment (martin-gorner) [reply packages/tasks/src/model-libraries.ts]: done

**Pull request merged**
